JEI Sortable Panel Integration and other improvements.

IMarvinTPA opened the pull request:
Add some internal sorting for Weapons, Armor, and Tools.
Add the ability to filter on ClassName.
Add the ability to filter on ToolClass and ArmorType.
Add the ability for the tree to have neighboring nodes to share the same sort index to allow other sorting to take place.
Integrate with JEI's customizable sortable item panel.
Add alternate API to unsort items that are not matched by the tree.
Updated Forge version.
Updated Item Tree XML for modded items.

For Issues #404, #555, #556

IMarvinTPA committed:
Add some internal sorting for Weapons, Armor, and Tools.
Add the ability to filter on ClassName.
Add the ability to filter on ToolClass and ArmorType.
Add the ability for the tree to have neighboring nodes to share the same sort index to allow other sorting to take place.
Integrate with JEI's customizable sortable item panel.
Add alternate API to unsort items that are not matched by the tree.
Updated Forge version.
Updated Item Tree XML for modded items.

Kobata commented on gradle.properties:
```diff
@@ -1,9 +1,9 @@
 mod_group=inventorytweaks
 mod_name=InventoryTweaks
-mod_version=1.64-dev
+mod_version=1.65-dev
```
I don't see a final 1.64 anywhere, so this is fairly unnecessary.

Kobata commented on src/main/java/invtweaks/InvTweaks.java:
```diff
@@ -448,16 +468,180 @@ int compareItems(@NotNull ItemStack i, @NotNull ItemStack j, int orderI, int ord
+        		int curDamage1 = i.getItemDamage();
+        		int curDamage2 = j.getItemDamage();
+        		if (curDamage2 != curDamage1) {
+                    if(i.isItemStackDamageable() && !getConfigManager().getConfig().getProperty(InvTweaksConfig.PROP_INVERT_TOOL_DAMAGE).equals(InvTweaksConfig.VALUE_TRUE)) {
```
There seem to be some general formatting inconsistencies around all this, especially in the preview this line is indented *less* than the one before it.

IMarvinTPA replied:
I've been working on JEI and it is all-tabs-all-the-time.  It seems InvTweaks is all-space-all-the-time.  So, I'll need to keep that in mind and work out a fix.

Kobata commented on src/main/java/invtweaks/InvTweaksItemTree.java:
```diff
@@ -4,13 +4,20 @@
 import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
+import net.minecraftforge.fml.common.registry.ForgeRegistries;
 import net.minecraftforge.oredict.OreDictionary;
+import scala.Int;
```
I think this is the wrong import for that. There isn't much reason for anything from `scala` to be imported.

(Added later: Also looks unused)

IMarvinTPA replied:
I had been using Int.MaxValue().  I've been swapping that out and forgot to remove the import.

Kobata commented on src/main/java/invtweaks/InvTweaksItemTree.java:
```diff
@@ -48,8 +55,13 @@
     private List<OreDictInfo> oresRegistered = new ArrayList<>();
+    
+    @NotNull
+    private List<ItemStack> allGameItems = new ArrayList<ItemStack>();
```
Nothing outside of here seems to ever use this, is there a reason for it to exist? It's potentially a good chunk of memory in highly-modded installs just sitting around persistently.

IMarvinTPA replied:
It now clears the list when the tree is done loading and (re)creates the list when the tree is reloaded.

Kobata commented on src/main/java/invtweaks/InvTweaksItemTreeLoader.java:
```diff
@@ -5,6 +5,8 @@
 import net.minecraft.nbt.NBTException;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraftforge.common.MinecraftForge;
+import scala.util.parsing.json.JSON;
```
Scala stuff again. Also appears entirely unused?

Kobata commented on src/main/resources/assets/inventorytweaks/itemtree.xml:
```diff
@@ -16,1181 +16,10191 @@
+		<weapon>
+			<meleeWeapon>
+				<sword>
+					<itemCreativeSword id="randomthings:creativesword"/>
```
The default item tree should probably not explicitly name any modded items. I do some general oredict categories and such to catch common ones but there's way too many popular mods (and too many items in them) to make a heavy attempt to keep up on them.

Perhaps a secondary optional file containing the expanded entries would be more suitable?

Kobata commented:
The travis issue is just it apparently being mad that PR checks don't get the signing key, I don't think there's been any (PRs) since that was added?

I pointed it out a little above but there's definitely something weird up with the formatting of all the changes: Most obviously around spaces vs tabs since that makes things align wrong on the web preview.

IMarvinTPA committed:
Attempting to track down a comparator contract violation.
Removed scala references.
Released items list after loading tree.

IMarvinTPA committed:
Fix tabs and spaces.
Add hook to refresh JEI's listing when tree is updated.

IMarvinTPA committed:
Fixed Comparator breaking the Comparator contract.
Added second option for JEI sorting for all items (mostly to test the non-api mode.)

IMarvinTPA committed:
Cleaned up imports and whitespace.

IMarvinTPA reviewed (commented):
General comment ended up in the wrong spot.

IMarvinTPA commented:
For the ItemTree, I've had hacked-up JEIs and InventoryTweaks that supported me in building the trees by seeing the whole tree at once in the inventory panel.  The current tree gets about 52% of all the items in StoneBlock.  I'm hoping by getting these integrations together, it'll free me up to do more tree maintenance rather than re-hacking.  I would hope that having a beefier tree, more people would be more comfortable with the inventory tweaks sorting and maybe even assist in improving it.

The formatting should be improved greatly.  I tried to get git filter processing working, but ultimately did a search-and-replace on tab to 4 spaces.

Thank You.

Kobata commented:
I've just been generally wary of adding too much setup specifically for certain mods.

Although, like I said I think there is a place for a more complete tree but that it should be separate from the vanilla-only one: both because there are a lot of people who don't really use many other mods with this, and because it makes for a much easier to update thing for new Minecraft versions, when a lot of mods end up massively changing their items and blocks anyway.

(The ideal would probably be a massive rework of the whole tree loading system so there can be different files merged but doing that obviously wouldn't be amazingly practical at the moment)

IMarvinTPA committed:
The mergePrevious attribute was not working as intended.  Fixed logic
with getting the order number when it is true.  Also, "==" string
comparisons were all false all the time.  Switched to .equals functions
and it works.

IMarvinTPA commented:
Can I add a button next to the "Edit Tree" that says "Load Default Modded Tree" that swaps them out?

Kobata commented:
Yeah I guess that'd work.

IMarvinTPA committed:
Added Shield detection.  Reverted and generically improved the default item tree.

IMarvinTPA committed:
Actually check in the updated vanilla tree.

IMarvinTPA committed:
Fix tabs to spaces.

IMarvinTPA committed:
Added a Modded Tree file.
Added a button on the settings panel to load the Modded Tree File.
Added debug output to inspect two items (compares each hand), disabled
with constant.

Kobata commented on src/main/resources/assets/inventorytweaks/lang/en_US.lang:
```diff
@@ -39,6 +39,8 @@ invtweaks.settings.rulesfile=Open the sorting rules file...
 invtweaks.settings.treefile=Open the item tree file...
 invtweaks.settings.treefile.error=[9] Failed to open tree file
+invtweaks.settings.moddedtreefile=Load modded tree file
+invtweaks.settings.moddedtreefile.error=Failed to replace tree file
```
This should probably be prefixed with a number: [14] looks like the first unused one (and wouldn't require all the translations to be updated) even though that breaks the pattern of going from top to bottom in the file

IMarvinTPA replied:
The actual number is 17 and the pattern is already broken.

Kobata commented on src/main/java/invtweaks/InvTweaksItemTreeLoader.java:
```diff
@@ -18,6 +19,7 @@
 import java.util.ArrayList;
 import java.util.LinkedList;
 import java.util.List;
+import java.util.Locale;
```
I don't see any uses of this import in the diff.

IMarvinTPA replied:
I removed ti from the toLowerCase when I was having troubles matching stringvariable == "constant".  Removed unused import.

IMarvinTPA committed:
ToolClass matching improved.

Make the tool class matching less greedy on the tree matching side for
more consistent results.
Fishing Rod added to the pseudo-tool list like shears and hoes.
Updated item trees.
Removed unused import.

Kobata commented:
Ok, is everything done enough to merge? I don't see anything remaining incredibly obvious.

IMarvinTPA commented:
I think I'm done with the features.  I'll send a new pull request when I have done a fair bit more to the moddeditemtree file.

Kobata commented:
Alright, shoved it in.
